File: java/08_006.java
import java.util.HashMap;
import java.util.Map;

public class Solution {
    public String optimalDocumentSegment(String document, 
                                        String tokens) {
        if (document == null || tokens == null || 
            document.isEmpty() || tokens.isEmpty()) {
            return "";
        }
        
        // Count frequencies of tokens
        Map<Character, Integer> tokenCount = new HashMap<>();
        for (char c : tokens.toCharArray()) {
            tokenCount.put(c, tokenCount.getOrDefault(c, 0) + 1);
        }
        
        // Initialize variables
        Map<Character, Integer> windowCount = new HashMap<>();
        int requiredTokens = tokenCount.size();
        int formed = 0;
        
        // Result tracking
        int minLen = Integer.MAX_VALUE;
        int resultStart = 0;
        
        // Window pointers
        int left = 0, right = 0;
        
        // Expand window
        while (right < document.length()) {
            // Add right character to window
            char c = document.charAt(right);
            windowCount.put(c, windowCount.getOrDefault(c, 0) + 1);
            
            // Check if this character satisfies a token requirement
            if (tokenCount.containsKey(c) && 
                windowCount.get(c).intValue() == 
                tokenCount.get(c).intValue()) {
                formed++;
            }
            
            // Try to contract window from left
            while (left <= right && formed == requiredTokens) {
                c = document.charAt(left);
                
                // Update result if current window is smaller
                if (right - left + 1 < minLen) {
                    minLen = right - left + 1;
                    resultStart = left;
                }
                
                // Remove left character from window
                windowCount.put(c, windowCount.get(c) - 1);
                
                // Check if removing this character breaks a token 
                // requirement
                if (tokenCount.containsKey(c) && 
                    windowCount.get(c) < tokenCount.get(c)) {
                    formed--;
                }
                
                left++;
            }
            
            right++;
        }
        
        return minLen == Integer.MAX_VALUE ? "" : 
            document.substring(resultStart, resultStart + minLen);
    }
}

File: data structures and algorithms/cs/08_006.cs
using System.Collections.Generic;

public class Solution {
    public string OptimalDocumentSegment(string document, string tokens) {
        if (string.IsNullOrEmpty(document) || string.IsNullOrEmpty(tokens)) {
            return "";
        }
        
        // Count frequencies of tokens
        Dictionary<char, int> tokenCount = new Dictionary<char, int>();
        foreach (char c in tokens) {
            if (tokenCount.ContainsKey(c)) {
                tokenCount[c]++;
            } else {
                tokenCount[c] = 1;
            }
        }
        
        // Initialize variables
        Dictionary<char, int> windowCount = new Dictionary<char, int>();
        int requiredTokens = tokenCount.Count;
        int formed = 0;
        
        // Result tracking
        int minLen = int.MaxValue;
        int resultStart = 0;
        
        // Window pointers
        int left = 0, right = 0;
        
        // Expand window
        while (right < document.Length) {
            // Add right character to window
            char c = document[right];
            if (windowCount.ContainsKey(c)) {
                windowCount[c]++;
            } else {
                windowCount[c] = 1;
            }
            
            // Check if this character satisfies a token requirement
            if (tokenCount.ContainsKey(c) && windowCount[c] == tokenCount[c]) {
                formed++;
            }
            
            // Try to contract window from left
            while (left <= right && formed == requiredTokens) {
                c = document[left];
                
                // Update result if current window is smaller
                if (right - left + 1 < minLen) {
                    minLen = right - left + 1;
                    resultStart = left;
                }
                
                // Remove left character from window
                windowCount[c]--;
                
                // Check if removing this character breaks a token requirement
                if (tokenCount.ContainsKey(c) && windowCount[c] < tokenCount[c]) {
                    formed--;
                }
                
                left++;
            }
            
            right++;
        }
        
        return minLen == int.MaxValue ? "" : document.Substring(resultStart, minLen);
    }
}